### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by Dadadon**
**Created models**
- Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
- Set up associations between models.
- author_id column in posts table is the foreign_key for the users table.

**Models include some additional methods.**

> Users
- A method that returns the 3 most recent posts for a given user.

> Posts
- A method that updates the posts counter for a user.
- A method that returns the 5 most recent comments for a given post.

> Comments
- A method that updates the comments counter for a post.

> Likes
- A method that updates the likes counter for a post.

### CODE REVIEW

**Review by Kamba56 — COMMENTED** [review]
Hi @Dadadon 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that **linters are failing.** Kindly fix linter errors.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app/models/like.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (FrederickMih) [inline app/models/like.rb]: - [ ] Please kindly fix this error. Remember we are using the author as seen in the Entity Relationship Diagram that was given to us

**belongs_to :author, class_name: 'User'**

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  has_many :comments
+  has_many :likes
+
+  validates :title, presence: true
+  validates :text, presence: true
+
+  def update_user_posts_counter
+    author.update(posts_counter: author.posts.count)
```
Comment (FrederickMih) [inline app/models/post.rb]: - [ ] Let's kindly model it this straightforward way: :+1: 

`author.increment!(:posts_counter)`

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post, counter_cache: true
+
+  def self.update_post_comments_counter(comment)
+    post = comment.post
+    post.update(comments_count: post.comments.count)
+  end
```
Comment (FrederickMih) [inline app/models/comment.rb]: - [ ] It's more easily doing it this way:

```
  def update_post_comments_counter(comment)
    post.increment!(:comments_counter)
  end
```

It seems you use comments_counter in the table not comments_count. Always double check columns in the db

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  belongs_to :post
+
+  def self.update_likes_counter(post_id)
+    post = Post.find(post_id)
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (FrederickMih) [inline app/models/like.rb]: - [ ] Kindly check the comments above :arrow_up:

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi :raising_hand:  @Dadadon, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!





## Highlights :green_circle: 

- PR is professional :+1: 
- Linter checks passed :heavy_check_mark: 


## Required Changes ♻️

- [ ] The license link of the readme file is broken. Kindly fix the link to render the readme file professional


_Check the comments under the review._

## Optional suggestions

- Please customize your text description or kindly remove it. This should done throughout the entire readme file. Screenshot example below:
Screenshot from 2023-03-02 14-40-53
- Kindly remove the deployment section since this app is not yet deployed
- Please update the **Future Features**  section of the readme file with the correct information
- kindly remove **FAQ (OPTIONAL)**  or fill in the section accordingly
- The text on the screenshot below should be removed in order to render your readme file professionally:
Screenshot from 2023-03-02 14-48-11





_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/like.rb (hunk 1 above)
Comment (Dadadon) [reply app/models/like.rb]: The ERD shows user_id not author @FrederickMih

Comment (FrederickMih) [reply app/models/like.rb]: > The ERD shows user_id not author @FrederickMih

Hi @Dadadon, you made that error in your `likes` table when surely generating migration for your L`ike `model. Kindly accept and correct it. Check your post table and kindly do the same.

File: app/models/like.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class Like < ApplicationRecord
+  # belongs_to :user
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  belongs_to :post
+
+  def self.update_likes_counter(post_id)
+    post = Post.find(post_id)
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (Ismailco) [inline app/models/like.rb]: - [ ] As the previos reviewer mentioned kindly make sure to change the `update_like_counter` method just like what you did in the `comment` model
```ruby
def update_likes_counter(post_id)
   post.increment!(:likes_counter)
end
```

**Review by Ismailco — CHANGES_REQUESTED** [review]
### Hi @Dadadon, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.


 ## Highlights

✔️ No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

 ## Required Changes ♻️
- [ ] Kindly make sure to include a `LICENSE` in your README file as the previous reviewer suggested so the license link is working
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by rotimiazeez — APPROVED** [review]
Hi @Dadadon 👏🏽, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlight
- Good job working on all the changes requested by the previous reviewers 👍🏽 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer, and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me(@rotimiazeez) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561), you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**